### ISSUE TITLE
Cannot update ComfyUI "as file is beyond symlink"

### ISSUE THREAD

**Issue by Torcelllo** [comfyanonymous/ComfyUI#1996]
Little as I know what causes this, but I think that an accompanying YAML file (for whatever model) has been missed during SymLinking the Models Folder.

Is there any workaround at all?

**Comment by NeedsMoar** [comment]
On Windows (I'm assuming that's where this is), python has a broken implementation of the function that determines what a symlink points to...  I think git is more or less broken when dealing with directory symlinks as well and only handles file symlinks.  I tracked down the exact problem in the python source a few months ago and then promptly forgot to file a bug on it because looking at python's source code is painful but IIRC it just checked whether the files resolved to the same directory without walking them correctly and also made an assumption about the size when checking that they were identical (windows symlinks are about 1kb, not 0 like the directory they point at) when it needed to be comparing the inode # of the target and the link file to get a correct answer.  Junctions worked but aren't supposed to be used for anything that doesn't have a shell handler (the user libraries mainly) so I wouldn't do that either, but what you're meant to do is edit:

extra_model_paths.yaml.example

and change
```yaml
    comfyui:
           base_path:  Z:\ComfyModelRoot\
```
To point to the location of your \models\ and rename it to extra_model_paths.yaml.  I'd try putting the models somewhere that's not a symlink, remove the symlink to wherever you had them so it won't interfere with git pull and try it again.   If you have other webUIs you can map models from them in that file too.

**Comment by Torcelllo** [comment]
I just deleted a Symlink which was delaying the process!

**Comment by polarNerd204** [comment]
Im on mac and have a simulated link to the models folder to store all the models on external drive. I just added a z_ infront of the models folder name then made new models folder ran update it worked.....now just go back and putt zz_ in front of new models folder and then remove the z_ from the simulated link one. Updated and back to normal.
 hope that helps someone else.

**Comment by DarkAlchy** [comment]
Linux here, and it is some seriously whacked nonsense.  stderr: 'error: 'models/checkpoints/put_checkpoints_here' is beyond a symbolic link because I have the models folder pointed (symlink) to a different location to unify and not waste space.  PITA to delete the symlink, update, then add it back.  I do have it in the extra model's path yaml but when I run Comfy the folder is recreated and it downloads other models to put back in there ignoring the yaml.

**Comment by ltdrdata** [comment]
Basically, the update system is managed by Git, and replacing a path with a symbolic link is considered a file change by Git. Avoid using the method of replacing the models directory with a symbolic link. That’s precisely why the `extra_model_paths.yaml` file is provided.

**Comment by DarkAlchy** [comment]
As I said, it doesn't work, and I am unsure why.  I delete the symlink and restart comfy to have it refilled with 1.8 gb of models.  Once that is over, I can simply delete the models folder and replace with a symlink and carry on.  The whole point was that some nodes are obviously not playing nice with the .yaml model directive.

**Comment by KLL535** [comment]
> Basically, the update system is managed by Git, and replacing a path with a symbolic link is considered a file change by Git. Avoid using the method of replacing the models directory with a symbolic link. That’s precisely why the `extra_model_paths.yaml` file is provided.

extra_model_paths.yaml
Here is my file: there is not a word about models and paths to them. How to use it?
```
# Config for testing nodes
testing:
    custom_nodes: tests/inference/testing_nodes
```

**Comment by ltdrdata** [comment]
It has nothing to do with `.gitignore`. Since the `ComfyUI/models` path itself is part of the files tracked by the `ComfyUI` repository, any change to that directory is considered a modification of the repository.  

Typically, the updating script performs a `git stash` to restore the original state to prevent conflict before `git pull`

If you insist on using a symbolic link, the only solution is to create a custom script to handle that specific situation.

**Comment by DarkAlchy** [comment]
You would think the nodes would behave and use the path I set in the .yaml, but a few do not.  I think there are 4, or 5, that do not.

**Comment by DarkAlchy** [comment]
> > You would think the nodes would behave and use the path I set in the .yaml, but a few do not. I think there are 4, or 5, that do not.
> 
> I had to rewrite the code of my nodes after switching to extra_model_paths.yaml path change. Now the new path does not replace the old one in `folder_paths.get_folder_paths("loras")[0]`, but is added as another path to the list. And you need to go through the entire list and search for files on each of the paths.

These offending nodes have always done it, and why I switched to symlink to begin with.  Must be going on two years now (sure feels like it at least) and still the same nodes have been updated over and over to still ignore the .yaml.  What gets me is they will load from it, but when I update force the files so it doesn't read the yaml then.  Very strange behaviour.

**Comment by KLL535** [comment]
> > > You would think the nodes would behave and use the path I set in the .yaml, but a few do not. I think there are 4, or 5, that do not.
> > 
> > 
> > I had to rewrite the code of my nodes after switching to extra_model_paths.yaml path change. Now the new path does not replace the old one in `folder_paths.get_folder_paths("loras")[0]`, but is added as another path to the list. And you need to go through the entire list and search for files on each of the paths.
> 
> These offending nodes have always done it, and why I switched to symlink to begin with. Must be going on two years now (sure feels like it at least) and still the same nodes have been updated over and over to still ignore the .yaml. What gets me is they will load from it, but when I update force the files so it doesn't read the yaml then. Very strange behaviour.

Probably because I didn't find any basic description of how and what to do. For example, I didn't even know that links can be redirected via a yaml file, I always used symbolic links and nowhere was it written that updates (and git) wouldn't work that way.

**Comment by DarkAlchy** [comment]
Oh, I am not pointing fingers for a reason, and what you said is what I thought.  I just wish it was more common knowledge as we can't do symlinks due to git and having two sets of models is a crazy space hog.

**Comment by ltdrdata** [comment]
It is not possible to replace the model folder itself with a symlink, but it is possible to replace individual files with symlinks. Using this method, it is somewhat possible to mitigate the offensive model downloading of certain nodes.

A better approach would be to suggest to the developers of the node pack to determine the model downloading path based on the `folder_paths` module API.

**Comment by DarkAlchy** [comment]
They seem unresponsive, so not sure.  So many models the symlink to each would be insane, and I suppose I could write something to do it.

I, along with others on the net, don't get why symlinks to folders, while to files, is verboten in github.

**Comment by ltdrdata** [comment]
> They seem unresponsive, so not sure. So many models the symlink to each would be insane, and I suppose I could write something to do it.
> 
> I, along with others on the net, don't get why symlinks to folders is verboten while to files, in github, it is not?

That's not the case. Whether it's a file or a folder, a symlink is essentially treated as a file pointing to a specific location. 
If you were to use a hard link instead, it would be indistinguishable.

**Comment by DarkAlchy** [comment]
Hardlink just copies the file so back to the lost space since two of the same file exists.

**Comment by ltdrdata** [comment]
> Hardlink just copies the file so back to the lost space since two of the same file exists.

That's not correct. A hard link allows multiple paths to access the same data without creating data duplication. However, it is not recommended due to potential security issues.

**Comment by DarkAlchy** [comment]
I tried a hardlink, and it copied the file so my files were double and the space taken was now twice as much.

**Comment by ltdrdata** [comment]
> I tried a hardlink, and it copied the file so my files were double and the space taken was now twice as much.

That is a strange phenomenon. The key point of a hard link, like a symbolic link, is that when one side is modified, the changes are reflected on the other side as well. This cannot be achieved through copying.

**Comment by n0valis** [comment]
Sorry for just coming in with a question similar but a bit different.

I am on a WSL  and would like to refer my image output to my Windows Images folder on /mnt/d/Images.
I tried --output-directory /mnt/d/Images and I tried ln -s /mnt/d/Images Images in the output folder.
Both would lead to the GUI not loading. As soon as I remove the link or --output-directory from the start command it would load again.
So the reference to  /mnt/d or any other drive would not work. These are the windows system mounts that are auto established within the wsl distro. All permissions are drwxrwxrwx so that cannot be the problem. Also if I create /mnt/data/Images and use it as --output-director that would work as well.
Any idea why I cannot use the system mounts?
